From 27fb5e0eed08d1ce7e3f3f93bd4cd355199a440d Mon Sep 17 00:00:00 2001
From: Ken Zhu <guigenz@codeaurora.org>
Date: Tue, 2 Jan 2018 11:49:41 -0800
Subject: [PATCH] netfilter: Fix multicast flow is not applied SNAT correctly

When bridge_nf_call_iptable is enabled, multicast flow
can reach nf_nat_inet_fn twice for SNAT operation, the
first time, it floods/snoops all ports of the bridge,
it won't hit nat rule in the do_chain, so
no nat operation required and finalized the conntrack,
and the second time, it reach nf_nat_inet_fn for forwarding
bridge to an upstream port, no do_chain is performed and
using the first time's result.

This is not the expected behavior. If the multicast is
forwarding between ports of a bridge, accept it directly.

Change-Id: I3dd3be2bd718051292a46485a1863cf127af4866
Signed-off-by: Ken Zhu <guigenz@codeaurora.org>
---
 net/netfilter/nf_nat_core.c | 25 +++++++++++++++++++++++++
 1 file changed, 25 insertions(+)

--- a/net/netfilter/nf_nat_core.c
+++ b/net/netfilter/nf_nat_core.c
@@ -24,6 +24,7 @@
 #include <net/netfilter/nf_nat.h>
 #include <net/netfilter/nf_nat_helper.h>
 #include <uapi/linux/netfilter/nf_nat.h>
+#include <linux/netfilter_bridge.h>
 
 #include "nf_internals.h"
 
@@ -725,6 +726,30 @@ nf_nat_inet_fn(void *priv, struct sk_buf
 	case IP_CT_RELATED_REPLY:
 		/* Only ICMPs can be IP_CT_IS_REPLY.  Fallthrough */
 	case IP_CT_NEW:
+#if IS_ENABLED(CONFIG_BRIDGE_NETFILTER)
+		/* when skb is forwarding between ports of a bridge,the
+		 * nf_bridge will be set and nf_bridge->physoutdev is not null,
+		 * We can assume that it is not expecting NAT operation.
+		 * when BR_HOOK is enabled, multicast packets will reach
+		 * postrouting twice,the first time is when it is forwarded
+		 * between ports of a bridge, the second time is that it is
+		 * forwarded to upstream port.
+		 *
+		 * It will perform traversing of the NAT table at the first
+		 * time, the next time, it will use the result of first time.
+		 * since forwarding betweeng ports of a bridge, it won't hit
+		 * rules of SNAT, it cause NO NAT operation on this skb when
+		 * forwarding to the upstream port.
+		 * To avoid the scenario above, accept it when it is forwarding
+		 * between ports of a bridge for multicast.
+		 */
+		if (skb->pkt_type == PACKET_MULTICAST) {
+			struct nf_bridge_info *nf_bridge =
+				nf_bridge_info_get(skb);
+			if (nf_bridge && nf_bridge->physoutdev)
+				return NF_ACCEPT;
+		}
+#endif
 		/* Seen it before?  This can happen for loopback, retrans,
 		 * or local packets.
 		 */
